### PR TITLE
Improve goto-symbols popup

### LINKED ISSUE #1069: Add path to goto definition/declaration popup
At the moment if symbols of the same name are defined in identically named files, it's hard to distinguish which file is which because there's no path in the popup.

The popup should show part of the path until a directory where the paths differ so it's possible to distinguish the different files. At the same time there should probably be some top limit for the length of the paths as they can make the popup too wide.

### PR DESCRIPTION
**Pull request by kugel-**
Fixes #1069 by  implementing the suggestions there.

See commit messages for details. tl;dr: the goto-symbols popup will show more of the paths, but as little as possible. The common prefix is stripped and the longest common sub-path is ellipsized.

Example: Assume the popup would show utils.h twice (/home/kugel/geany.git/src/utils.h and /home/kugel/geany.git/build/dest/include/geany/utils.h.

The popup would show:
src/utils.h
build/dest/include/geany/utils.h

Additionally, as per @elextr suggestion and for a frequent use-case of mine, the ellipsis is introduced for long common substrings, which I often have due to having the same code base checked out multiple times (my workflow at work requires this).

So, /home/kugel/checkout_a/path/to/project/src/main.c and /home/kugel/checkout_b/path/to/project/src/main.c shows as:
checkout_a/.../main.c
checkout_b/.../main.c

Preview (/home/kugel is stripped of the front):
grafik

### CODE REVIEW

File: src/utils.c [hunk 1]
```diff
@@ -2031,6 +2031,199 @@ gchar **utils_strv_join(gchar **first, gchar **second)
...
+		return NULL;
+
+	if (num == 0)
+		num = g_strv_length(strv);
+
+	prefix = g_strdup(strv[0]);
+
+	for (gint i = 0; prefix[i]; i++)
+	{
+		foreach_strv(ptr, &strv[1])
```
Comment (codebrainz) [inline src/utils.c]: Guessing it's accidental since it's indented wrong and missing braces, and `strv` can be less than two elements long and not necessarily nul-terminated.

A bit off-topic but what does this weird macro do compared to the normal C code `for (GStrv ptr = strv+1; *ptr; ptr++)`?

**Comment by kugel-** [comment]
> can't figure out the G* macro shit.

which G* macro shit? `foreach_strv` is a Geany macro. If it's hated it should be deprecated and/or removed.

File: src/utils.c (hunk 1 above)
Comment (kugel-) [inline src/utils.c]: Err, this line should be simply deleted.My mistake.

File: src/utils.c [hunk 2]
```diff
@@ -2031,6 +2031,199 @@ gchar **utils_strv_join(gchar **first, gchar **second)
...
+/* * Returns the common prefix in a list of strings.
+ *
+ * The size of the list may be given explicitely automatically determined if passed a GStrv.
+ *
+ * @param strv The list of strings to process.
+ * @param num The number of strings contained in @a strv. Can be 0 if @a strv is a @c GStrv
+ *
+ * @return The common prefix that is part of all strings.
+ */
+gchar *utils_strv_find_lcs(gchar **strv, size_t num)
```
Comment (kugel-) [inline src/utils.c]: Oops, I coped the docs but then didn't change it. Will fix.

File: src/utils.c [hunk 3]
```diff
@@ -2031,6 +2031,199 @@ gchar **utils_strv_join(gchar **first, gchar **second)
 	return strv;
 }
 
+/* * Returns the common prefix in a list of strings.
+ *
+ * The size of the list may be given explicitely automatically determined if passed a GStrv.
```
Comment (kugel-) [inline src/utils.c]: I was under the impression that GStrv also implied NULL termination since it's expected by `g_strv_length` and `g_strv_contains`. I will change the comment.

**Comment by elextr** [comment]
@kugel- to be clear, nothing I said is a general criticism of the purpose your PR or the implementation (apart from the specific comments you have already fixed).  

I just found it difficult to understand it, not aided by lower case macros that "capture" following syntax.  Also as you pointed out above, not only is it lower case, it is an unprefixed Geany local macro.  So I was not confident I understood the implementation, and couldn't justify spending more time on it.  Thats why I approved my review, so someone else could review it and not need to override my review.

Yes I believe macros like foreach_strv and foreach_document etc should be removed, but thats not part of this PR, will open another issue for discussions about that.

**Comment by kugel-** [comment]
@elextr @b4n @codebrainz Can someone please review this?

**Comment by kugel-** [comment]
@elextr @b4n @codebrainz ping

**Comment by vfaronov** [comment]
I gave this PR a brief spin. Linux and Windows, GTK+2 and GTK+3, a few realistic and contrived examples.

-----------------------------------------------

**Bug:** on Windows, if one of the pathnames to be rendered in the popup is too long, it is considered empty. This happens before the removal of common prefixes, so it looks like this:

geany-goto-popup-cut

This first line “: 1” actually refers to a very long pathname. If I click it, Geany successfully jumps to the symbol in that pathname, but also prints this to the console:

    Geany: utils_tidy_path: assertion 'g_path_is_absolute(filename)' failed

So what is a “very long pathname”? Here is an example:

    C:\Users\Vasiliy\Downloads\Первая папка\В ней много интересного\Приколов не счесть\Всякое разное 1234567890\Ещё больше 1234567890\Последний уровень вло\Ор выше гор.c

As you can see, it’s 165 characters long, but in UTF-8 it would be 261 bytes long. Now this is the minimum length where the problem happens. If I remove any **1 Cyrillic** or any **2 Latin** characters from anywhere in this pathname, it is rendered as expected.

I could not reproduce this problem with ASCII-only pathnames, probably because Windows prevented me from creating a pathname longer than 255 characters.

I could not reproduce this problem on Linux.

-----------------------------------------------

When common substrings are ellipsized, Geany now prints messages to the console, like this:

```
p 0 Вторая папка/.../Тест.c
p 1 Первая папка/.../Тест.c
```

I think this should be either removed or made more understandable.

**Comment by kugel-** [comment]
@vfaronov Many thanks for testing.

Regarding the bug. I assume PATH_MAX is just 260 on Windows. I've read http://insanecoding.blogspot.de/2007/11/pathmax-simply-isnt.html and it appears PATH_MAX is pretty broken in general. 

The tag data contains the full path name. Apparently it contains something weird (non-null) if it exceeds PATH_MAX, while from the code I'd think it contains NULL (but that would crash Geany). To be sure, please check with the following diff.

In any case, this is a separate issue. We have our own `realpath()` on windows which doesn't handle this case. The the underlying data this PR uses is already wrong. You should see the error message about `utils_tidy_path()` even without my patch.

```
diff --git a/src/symbols.c b/src/symbols.c
index f4f259b72..874666399 100644
--- a/src/symbols.c
+++ b/src/symbols.c
@@ -1933,8 +1933,12 @@ static void show_goto_popup(GeanyDocument *doc, GPtrArray *tags, gboolean have_b
 	/* If popup would show multiple files presend a smart file list that allows
 	 * to easily distinguish the files while avoiding the file paths in their entirety */
 	file_names = g_new(gchar *, tags->len);
+	int n = g_random_int()%100;
 	foreach_ptr_array(tmtag, i, tags)
+	{
 		file_names[i] = tmtag->file->file_name;
+		printf("goto_popup %d: %s", n, tmtag->file->file_name ? tmtag->file->file_name : NULL);
+	}
 	short_names = utils_strv_shorten_file_list(file_names, tags->len);
 	g_free(file_names);
 
```

**Comment by vfaronov** [comment]
> To be sure, please check with the following diff.

I get an empty string for the path (`goto_popup 24: `).

> You should see the error message about `utils_tidy_path()` even without my patch.

I do. In fact, I can’t even open two different files with broken paths like that, as if Geany considers them to be the same file.

**Comment by kugel-** [comment]
@vfaronov Can you please open a new issue for this? Maybe we can find a solution indepedently.

**Comment by vfaronov** [comment]
@kugel- [Done.](https://github.com/geany/geany/issues/1534)

**Comment by kugel-** [comment]
@elextr @b4n @codebrainz please review and merge.

**Comment by kugel-** [comment]
@elextr @b4n @codebrainz ping

File: src/symbols.c [hunk 4]
```diff
@@ -1927,15 +1927,23 @@ static void show_goto_popup(GeanyDocument *doc, GPtrArray *tags, gboolean have_b
 	GdkEventButton *button_event = NULL;
 	TMTag *tmtag;
 	guint i;
-
+	gchar **short_names, **file_names, **p;
 	menu = gtk_menu_new();
 
+	/* If popup would show multiple files presend a smart file list that allows
```
Comment (elextr) [inline src/symbols.c]: `presend` sb `present`?

**Comment by elextr** [comment]
@kugel- I got the same reaction looking at it again as last time, so I am never gonna "review" it, and I don't use the gotos much so it isn't going to get tested much here either,  but given that @vfaronov has tested, it I might be willing to commit it so it can be tested by others.  The only question is for windows, can it at least show the filename if the path is too long, not just `: 1`, then at least its not worse than it is now :)

**Comment by kugel-** [comment]
@elextr For windows, the issue has nothing to do with my change. The tags contain the wrong path (or no path), even without my change. The display issue is present in master, too.

**Comment by elextr** [comment]
@kugel- ok, if the windows problem is an existing issue, then if nobody objects loudly over the next week, then I will be willing to commit it if you fix the conflicts.

**Comment by kugel-** [comment]
@elextr the only conflict is about the API version increment. I thought it was generally practice that this kind of conflict is handled upon merge (since otherwise PRs have to be constantly updated against API version changes)

**Comment by b4n** [comment]
I'm in the process of reviewing this, and think that maybe the code could be a little simpler.  I'll post a review and suggestions in the upcoming days.

**Comment by elextr** [comment]
@b4n JUST sneaked inside the week, sorry @kugel- no commit.

**Comment by kugel-** [comment]
Am 16. September 2017 11:10:42 MESZ schrieb elextr <notifications@github.com>:
>@b4n JUST sneaked inside the week, sorry @kugel- no commit.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/geany/geany/pull/1445#issuecomment-329956618

No problem. An in-depth review is always welcome.

**Comment by kugel-** [comment]
@b4n What's your status?

**Comment by kugel-** [comment]
@b4n ping

**Comment by kugel-** [comment]
@b4n ping

**Comment by kugel-** [comment]
@b4n ping

**Comment by kugel-** [comment]
@b4n ping again...you are active on many other PRs, yet you ignore this one even though you said promised to review. Makes me sad.

**Comment by b4n** [comment]
I'm sorry. I don't have much time lately, and the reason I'm active one other ones and this one is unfortunately that this one is trickier and thus requires more time commitment than more trivial issues.
To try and get it moving again even with limited time, I'll drop some alternative code I wrote a while back. I wanted to post it integrated and with proper rationale, but well, you're still waiting. So just dropping it is probably better than nothing.

https://gist.github.com/b4n/0f715c19239f501200cfeaefa5a6979c
* IMO (at time of writing I thought it was, at least) easier to understand
* Probably more resources-greedy (likely more memory, likely more CPU), but I don't think it matters at this level
* Fixes quite a few bugs in shortening.  I unfortunately forgot what they all were and would need to find them all again, but IIRC the implementation proposed in the PR can count and/or remove sequences not entirely composed of full directories (meaning it could remove half a directory, or nothing at all when it should, with specially crafted path lists sharing longer substrings outside of directory portions or alike). You probably can try the same kind of things I did using a similar sample program as the one in my proposed implementation.

**Comment by kugel-** [comment]
@b4n to be honest, I prefer my version. I find it easier to grasp because it's split into 3 strv-related utility functions that have well defined and separated functionality (the functions should be useful on their own). I don't so how your code is really any shorter (and neither is worryingly long, each is well below 200 LOCs with 20-30 of them being comments). And it's more efficient. I'm surprised, as you usually seem care about efficiency? Efficiency is normally not a big deal for me, but I care if the more efficient version is already written (or if the change is trivial and/or obvious but that doesn't apply here).

So I would rather fix the bugs you mention and proceed with my version. What do you think?

**Comment by b4n** [comment]
Well, yeah I don't like inefficient code when it can easily be improved in that aspect.  OTOH, I prefer easy to read/maintain code.

And I'll have to check again, maybe I was tired, but I found your code oddly hard to follow, and tricky to fix the issues I found. At the time I found that working on path components directly made everything simpler at many level.

And for the matter of otherwise useful reusable functions, yes, but for example I think I remember one bug is that the longest common substring is *not* necessarily what you want here, like in */a/b/c/d/longilename* vs. */z/b/c/y/longfilename*: *longfilename* is obviously the longest substring, but you actually want */b/c/* -- so IIRC your code will just not shorten this because it'll find *longfilename* as the substring, but will rightly ignore it because it the basename.  So to fix this the function should be updated to find the longest substring *surrounded by separators*, becoming a lot less useful for other callers.

However yes, if you can fix the issues I'll try and give it a new shot and propose some changes instead of a totally different approach.

File: src/utils.c [hunk 5]
```diff
@@ -2031,6 +2031,198 @@ gchar **utils_strv_join(gchar **first, gchar **second)
...
+ *
+ * The algorthm strips the common prefix (e-g. the user's home directory) and
+ * replaces the longest common substring with an ellipsis ("...").
+ *
+ * @param file_names @array{length=num} The list of strings to process.
+ * @param num The number of strings contained in @a file_names. Can be 0 if it's terminated by @c NULL.
+ * @return @transfer{full} A newly-allocated array of transformed paths strings, terminated by
+            @c NULL. Use @c g_strfreev() to free it.
+ *
+ * @since 1.31 (API 232)
```
[the remark this answers is not in the document]
Comment (kugel-) [reply src/utils.c]: I need this function in peasy right away. I thought I mentioned that, didn't I?

**Comment by kugel-** [comment]
I hope the code is clearer now and the bug you mentioned should be fixed

**Comment by b4n** [comment]
Looking at the table below, is this expected, or is it a bug?

path | current result | what I expected
---- | ---- | ----
`/aaa/bbb/cc/ccccc/ddddd` | `cc/.../ddddd` | `.../ddddd`
`/aaa/bbb/xxx/yyy/cc/ccccc/ddddd` | `xxx/yyy/cc/.../ddddd` | `xxx/yyy/.../ddddd`

**Comment by b4n** [comment]
Also, your code doesn't seem to support Windows path separators.  I'm not 100% sure we need it in Geany here, but it would probably be good to have support for it if it's supposed to be a generic utility.

**Comment by kugel-** [comment]
> Looking at the table below, is this expected, or is it a bug?
it's debatable. From the plain algorithm, your expectation is right and the result is buggy. But one can argue that the actual result may be less confusing to the user (ellipsizing is clearer when the 3 dots are encosed by words. I've never seen a ellipsizing that cuts of the entire front).

> Also, your code doesn't seem to support Windows path separators
What do you mean? I used G_DIR_SEPARATOR everywhere. I didn't test on Windows but it should work.

**Comment by kugel-** [comment]
@b4n friendly ping :-)

**Comment by elextr** [comment]
> (ellipsizing is clearer when the 3 dots are encosed by words. I've never seen a ellipsizing that cuts of the entire front).

I agree with @kugel- here, and would have expected what @b4n showed as "current result" anyway.

**Comment by b4n** [comment]
@kugel- there still are problems shortening some paths, like e.g. */src/a/app-1.2.3/src/lib/module/source.c* and */src/b/app-2.2.3/src/module/source.c*:

path | current result | my expectation
------|--------------------|------------------
`/src/a/app-1.2.3/src/lib/module/source.c` | `a/app-1.2.3/src/lib/module/source.c` | `a/app-1.2.3/src/lib/.../source.c`
`/src/b/app-2.2.3/src/module/source.c` | `b/app-2.2.3/src/module/source.c` | `b/app-2.2.3/src/.../source.c`

**Comment by kugel-** [comment]
>
 > path     current result     my expectation
 > |/src/a/app-1.2.3/src/lib/module/source.c|
 > |a/app-1.2.3/src/lib/module/source.c| |a/app-1.2.3/src/lib/.../source.c|
 > |/src/b/app-2.2.3/src/module/source.c|
 > |b/app-2.2.3/src/module/source.c| |b/app-2.2.3/src/.../source.c|
 >

Damn. What happens is that the code determines `.2.3/src/` to be the 
longest common substring (longer that the expected `/module/`). Then
it reduces it to directory components, giving only `src`. This
is shorter than the threshold of 5 chars, so in the end nothing is 
ellipsized. It works if you change /module/ to /moduleXX/.

Fixing this isn't trivial, and without proper unit tests there is a risk 
to break previously working cases.

I suggest to accept the behavior for now (and merge) and I'll follow up 
with a fix + unit tests. But I would hate if this doesn't make it into 
the release just because of select (edge) cases because the status quo 
is just plain unusable for me. In fact I've been using this patch at 
work for ages where I have real-world cases and I never came across 
issues like this.

Thanks for looking into it again!

**Comment by b4n** [comment]
> Fixing this isn't trivial

That's why I suggested in an earlier comment that the fact `utils_strv_find_lcs()` looks generic is cute, but in practice this function does *not* do what `utils_strv_shorten_file_list()` needs.

> and without proper unit tests there is a risk to break previously working cases.

Yeah, we should probably add tests for this.  Fortunately, it's probably one of the few things that should be easy enough to add a test for, because `utils.c` don't usually have dependencies on internal Geany state, which means it can likely be tested by a minimal test program linking libgeany.

> I suggest to accept the behavior for now (and merge) and I'll follow up with a fix + unit tests. But I would hate if this doesn't make it into the release just because of select (edge) cases because the status quo is just plain unusable for me. In fact I've been using this patch at work for ages where I have real-world cases and I never came across issues like this.

Fair enough; I must admit I specifically crafted the test to fail, looking at the weakness of the implementation :)

This said, note that I have another implementation (see a message above) that doesn't have the issue.  It has arguably worse performance, and can probably use some improvement in other area, but I didn't find a bug in its results yet.

File: src/symbols.c (hunk 4 above)
Comment (b4n) [reply src/symbols.c]: @elextr seems to indeed have spotted a small typo :)

File: src/utils.c [hunk 6]
```diff
@@ -2031,6 +2031,209 @@ gchar **utils_strv_join(gchar **first, gchar **second)
...
+ * This function takes a list of file names (probably with absolute paths), and
+ * transforms the paths such that they are short but still unique. This is intended
+ * for dialogs which present the file list to the user, where the base name may result
+ * in duplicates (showing the full path might be inappropriate).
+ *
+ * The algorthm strips the common prefix (e-g. the user's home directory) and
+ * replaces the longest common substring with an ellipsis ("...").
+ *
+ * @param file_names @array{length=num} The list of strings to process.
+ * @param num The number of strings contained in @a file_names. Can be 0 if it's terminated by @c NULL.
```
[the remark this answers is not in the document]
Comment (kugel-) [reply src/utils.c]: g_malloc() doesn't accept 0 (returns NULL) so I don't think such 0-sized array (0 elemets and not NULL-terminated) can ever exist.

File: src/utils.c [hunk 7]
```diff
@@ -2031,6 +2031,209 @@ gchar **utils_strv_join(gchar **first, gchar **second)
 	return strv;
 }
 
+/* * Returns the common prefix in a list of strings.
+ *
+ * The size of the list may be given explicitely, but defaults to @c g_strv_length(strv).
+ *
+ * @param strv The list of strings to process.
+ * @param num The number of strings contained in @a strv. Can be 0 if it's terminated by @c NULL.
```
[the remark this answers is not in the document]
Comment (kugel-) [reply src/utils.c]: I can change if you prefer. But I couldn't find an example in the glib documentation.

File: src/utils.c [hunk 8]
```diff
@@ -2031,6 +2031,209 @@ gchar **utils_strv_join(gchar **first, gchar **second)
...
+	names = g_new0(gchar *, num + 1);
+
+	prefix = utils_strv_find_common_prefix(file_names, num);
+	/* First: determine the common prefix, that will be stripped.
+	 * Don't strip single-letter prefixes, such as '/' */
+	prefix_len = 0;
+	if (NZV(prefix) && prefix[1])
+	{
+		/* Only strip directory components, include trailing '/' */
+		start = strrchr(prefix, G_DIR_SEPARATOR);
```
[the remark this answers is not in the document]
Comment (kugel-) [reply src/utils.c]: Meh, what's the point G_DIR_SEPARATOR if you need to handle both. Also, what about legitimate paths that contain the other slash (e.g. a file can contain the \ character on linux, not sure if the same is true for / on Windows).

Are you sure we don't have problems with this elsewhere inside Geany? I.e. is this something we support or not?

[the remark this answers is not in the document]
Comment (b4n) [reply src/utils.c]: > e.g. a file can contain the \ character on linux

Yes, handling of `\` has to be special-cased for Windows

> not sure if the same is true for / on Windows

I don't think so, IIUC both `\` and `/` are separators.  One simple non-optimized way to match it can be `"/" G_DIR_SEPARATOR_S` (e.g. preprocessor concatenation of either `"//"` or `"/\"`) together with things like `strpbrk()`.

> Are you sure we don't have problems with this elsewhere inside Geany? I.e. is this something we support or not?

I'm not entirely sure, but I think so.  However, we're probably normalizing the path at some point, which means we might not encounter the issue with paths we handle anyway.
So this might not be so important in the end.

File: src/utils.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (kugel-) [reply src/utils.c]: I can change it, no problem. FWIW: I'd argue that passing NULL should be an error (might add an g_return_val_if_fail() since it's an API function).

[the remark this answers is not in the document]
Comment (b4n) [reply src/utils.c]: IMO it should not be an error if you allow a non-`NULL`-terminated array, because as you said yourself, allocating a 0-length array will give you a `NULL` pointer -- which is fine as you shouldn't be accessing anything past its 0th element.

[the remark this answers is not in the document]
Comment (codebrainz) [reply src/utils.c]: Probably a stupid question (I didn't look at the calling code), but couldn't it use a ready-made collection type like GArray or GPtrArray instead of passing separate arguments? If the parameter isn't modified and the caller does have separate pointer and length, they could just stuff them in a GArray on the stack and pass by const pointer. Or the function could mutate the array in place for better performance. Would also allow to deduplicate the input array without handing back an array that has to be counted by walking through it. Seems cleaner and more efficient than using a plain C array and separate length, but maybe I'm brainwashed by C++ by now, so feel free to ignore :)

[the remark this answers is not in the document]
Comment (kugel-) [reply src/utils.c]: strv is very gir friendly, and since the function would decompose a GPtrArray to handle its raw strv it might as well get the strv directly.

File: src/utils.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (kugel-) [reply src/utils.c]: >  So this might not be so important in the end.

Ok, so I don't plan changes for now in this regard. Speak up if you chnage your mind.

File: src/utils.c [hunk 9]
```diff
@@ -2031,6 +2031,208 @@ gchar **utils_strv_join(gchar **first, gchar **second)
...
+ *
+ * @param strv The list of strings to process.
+ * @param num The number of strings contained in @a strv. Can be -1 if it's terminated by @c NULL.
+ *
+ * @return The common prefix that is part of all strings (maybe empty), or NULL if an empty list
+ *         was passed in.
+ */
+static gchar *utils_strv_find_common_prefix(gchar **strv, gssize num)
+{
+	gchar *prefix;
```
Comment (b4n) [inline src/utils.c]: unused now

**Comment by b4n** [comment]
```
(geany:25173): Geany-CRITICAL **: utils_strv_shorten_file_list: assertion 'num != 0 && file_names == NULL' failed
Erreur de segmentation
```
oops!  Just tried to go to a location

File: src/utils.c [hunk 10]
```diff
@@ -2031,6 +2031,208 @@ gchar **utils_strv_join(gchar **first, gchar **second)
...
+gchar **utils_strv_shorten_file_list(gchar **file_names, gssize num)
+{
+	gsize i;
+	gchar *prefix, *substring, *lcs;
+	gchar *start, *end;
+	gchar **names;
+	gsize prefix_len, lcs_len;
+
+	/* We don't dereference file_names if num == 0. */
+	g_return_val_if_fail(num != 0 && file_names == NULL, NULL);
```
Comment (b4n) [inline src/utils.c]: This should be `num == 0 || file_names != NULL`, as it's basically an assertion, and should list conditions expected to pass.

**Comment by b4n** [comment]
@kugel- just in case you didn't see my review here like on the other PR.

**Comment by kugel-** [comment]
I rebased and imply removed the API bump that was causing the conflict. I hope that's what you wanted?

**Comment by b4n** [comment]
@kugel- See https://github.com/kugel-/geany/pull/6

**Comment by kugel-** [comment]
@b4n are we good to go?

**Comment by b4n** [comment]
Yep, thanks :)

**Comment by kugel-** [comment]
Awesome, thanks!
